crawl-ref/source/menu.cc

```diff
@@ -97,16 +97,181 @@ int Popup::pop()
 }
 #endif
 
-int MenuDisplayText::get_maxpagesize()
+class MenuDisplay
+{
+public:
+    MenuDisplay(Menu *menu) : m_menu(menu) {};
+    virtual ~MenuDisplay() {}
+    virtual void draw_stock_item(int index, const MenuEntry *me) = 0;
+    virtual void draw_items() = 0;
+    virtual bool item_in_page(int index) = 0;
+    virtual void set_offset(int lines) = 0;
+    virtual void draw_more() = 0;
+    virtual void set_num_columns(int columns) = 0;
+    virtual bool scroll_page(int delta) = 0;
+    virtual bool scroll_to_item(int index) = 0;
+    virtual void get_visible_items_range(int &first, int &last) = 0;
+protected:
+    Menu *m_menu;
+};
+
+class MenuDisplayText : public MenuDisplay
```

Comment:
Why move/put these two class definitions in the cpp file?

Reply:
Mostly because I saw these as implementation details of the menu, rather than interfaces to the menu's functionality